### ISSUE TITLE
text_box: abort when type many character by input method

### ISSUE THREAD

**Issue by fujimo-t** [wesnoth/wesnoth#2619]
To reproduce:
1. Push 'Multiplayer" in title screen
2. Click text box for login name
3. Activate input method
4. Push any key repeatly

My environment:
* OS: openSUSE Tumbleweed
* Input method framework: fcitx
* Input method engine: mozc
* wesnoth version: 1.13.11+dev (471c9e876a)

**Comment by CelticMinstrel** [comment]
Off the top of my head, I wonder whether this could actually be an upstream (SDL) bug. If I recall correctly, `TEXT_EDITING` events only have limited space in which to store the text being edited.

**Comment by fujimo-t** [comment]
Messages when abort:
```
wesnoth: /home/user/wesnoth/src/gui/widgets/text_box_base.cpp:146: void gui2::text_box_base::set_cursor(size_t, bool): Assertion `offset <= text_.get_length()' failed.
fish: “./wesnoth” terminated by signal SIGABRT (Abort)
```

Backtrace:
https://gist.github.com/fujimo-t/ad85412930636b778dec92ac7f3b07ad

**Comment by Vultraz** [comment]
That login name textbox has a max size of 20 characters.

**Comment by CelticMinstrel** [comment]
Well, that does suggest it's not an upstream bug, at least. The message suggests that the insertion point is out of bounds, I believe. I wonder if this is some quirk in SDL's handling of the event? For example, if it "scrolls" the view of the text being edited so that the insertion point is always within the text passed to `TEXT_EDITING`, then I could imagine this occurring... but then what would it do if you selected a portion that's longer than it can store in the event?

All this is just unsubstantiated theorizing though.

@Vultraz - You do need to use a Japanese or similar input method, according to the opening post. Maybe you're not doing that?

**Comment by Vultraz** [comment]
Ah, indeed I am not. I assume Japanese characters count as more than 1 Latin character? Perhaps that has something to do with it.

**Comment by CelticMinstrel** [comment]
No, if that were the issue then this would also affect, say, Greek. (Though I'm only _assuming_ it doesn't based on the description, I haven't actually tested.) Input method editors are significantly more complicated than normal typing; a good comparison would be the suggestions provided by your IDE when coding. In effect, an IME is a separate piece of software (often built into the OS) which intercepts the keyboard input before it reaches the application. Only when the user has "confirmed" their input does the application receive the final result; however, it's still kept updated as the user edits their input, so that it can draw what they're working with onscreen.

**Comment by fujimo-t** [comment]
I confirmed this problem on Windows 10.
Error messages and result of backtrace are same to Linux.

**Comment by Vultraz** [comment]
Fixed by #2644.
